Sources/Benchmark/BenchmarkState.swift

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Benchmark state is used to collect the 
+/// benchmark measurements and view the settings it 
+/// was configured with.
+/// 
+/// Apart from the standard benchmark loop, you can also use
+/// it with customized benchmark measurement sections via either
+/// `start`/`end` or `measure` functions.
+public struct BenchmarkState {
+    let iterations: Int
+    var startTime: UInt64
+    var endTime: UInt64
+    var measurements: [Double]
+
+    /// Aggregated settings for the current benchmark run. 
+    public let settings: BenchmarkSettings
+
+    @inline(__always)
+    init(iterations: Int, settings: BenchmarkSettings) {
+        self.iterations = iterations
+        self.startTime = 0
+        self.endTime = 0
+        self.measurements = []
+        self.measurements.reserveCapacity(iterations)
+        self.settings = settings
+    }
+
+    /// Explicitly marks the start of the measurement section.
+    @inline(__always)
+    public mutating func start() {
+        self.endTime = 0
+        self.startTime = now()
+    }
+
+    /// Explicitly marks the end of the measurement section
+    /// and records the time since start of the benchmark.
+    @inline(__always)
+    public mutating func end() throws {
+        let value = now()
+        if self.endTime == 0 {
+            self.endTime = value
+            try record()
+        }
+    }
+
+    @inline(__always)
+    mutating func record() throws {
+        if measurements.count < iterations {
+            measurements.append(self.duration)
+        } else {
+            throw BenchmarkTermination()
+        }
+    }
+
+    @inline(__always)
+    var duration: Double {
+        return Double(self.endTime - self.startTime)
+    }
+
+    @inline(__always)
+    mutating func loop(_ benchmark: AnyBenchmark) throws {
+        while true {
+            benchmark.setUp()
+            start()
+            try benchmark.run(&self)
+            try end()
+            benchmark.tearDown()
+        }
+    }
+
+    /// Run the closure within within benchmark measurement section.
+    /// It may throw errors to propagate early termination to 
```

Comment:
Presumably "it" refers to “the closure” which presumably refers to `f`.  If not the meaning of the statement is very confusing.  But if so, the statement lies.  Clearly `f` can't throw as it isn't marked `throws`.  The only thing that throws here is `end()`, which if I trace it back is propagating a throw from record on line 64.  And the logic behind what generates the error is a bit mysterious.  If you have as many measurements as iterations the process is supposed to stop?  Why?